### PR TITLE
Optionally don't inline stream annotated kCalls

### PR DESCRIPTION
**Pull request by chaserileyroberts**
First part of splitting up #17982

### CODE REVIEW

**Review by golechwierowicz — COMMENTED** [review]
Thank you for the PR split! Much appreciated!

File: xla/service/call_inliner_test.cc [hunk 1]
```diff
@@ -489,5 +489,33 @@ TEST_F(CallInlinerTest, UseShardManualComputationBodyInlined) {
...
+  auto debug_options = HloTestBase::GetDebugOptionsForTest();
+  debug_options.set_xla_gpu_experimental_stream_annotation(true);
+  auto module = ParseAndReturnVerifiedModule(hlo_string).value();
+  module->mutable_config().set_debug_options(debug_options);
+  CallInliner call_inliner(/*single_call_site=*/true);
+
+  TF_ASSERT_OK_AND_ASSIGN(bool mutated, call_inliner.Run(module.get()));
+  ASSERT_FALSE(mutated);
+
+  ASSERT_EQ(module->entry_computation()->instruction_count(), 2);
```
[the remark this answers is not in the document]
Comment (chaserileyroberts) [reply xla/service/call_inliner_test.cc:513]: Sure! I'll add a filecheck test and I'll also have this test inline a separate kCall in the graph that doesn't have the annotation.

**Comment by chaserileyroberts** [comment]
I've updated the test.

File: xla/service/call_inliner.cc [hunk 2]
```diff
@@ -160,6 +161,20 @@ bool InlineComposites(
              instruction->frontend_attributes().map().at("composite.name"));
 }
 
+bool InlineStreamAnnotation(HloInstruction* instruction) {
+  if (instruction->GetModule()
+          ->config()
+          .debug_options()
+          .xla_gpu_experimental_stream_annotation()) {
+    auto it =
```
Comment (sanjoy) [inline xla/service/call_inliner.cc:169]: Nit: use `ContainsKey`?

File: xla/xla.proto [hunk 3]
```diff
@@ -819,6 +819,8 @@ message DebugOptions {
   // Whether to use multiple compute streams to run windowed einsum.
   bool xla_gpu_multi_streamed_windowed_einsum = 280;
 
+  bool xla_gpu_experimental_stream_annotation = 341;
```
Comment (sanjoy) [inline xla/xla.proto:822]: Nit: add a comment describing what this does.

**Review by sanjoy — COMMENTED** [review]
Left some minor nitpicky comments inline.

File: xla/xla.proto (hunk 3 above)
Comment (chaserileyroberts) [reply xla/xla.proto]: Fixed.

File: xla/service/call_inliner.cc (hunk 2 above)
Comment (chaserileyroberts) [reply xla/service/call_inliner.cc]: Fixed.

**Review by golechwierowicz — APPROVED** [review]

File: xla/service/call_inliner_test.cc [hunk 4]
```diff
@@ -489,5 +490,58 @@ TEST_F(CallInlinerTest, UseShardManualComputationBodyInlined) {
...
+  TF_ASSERT_OK_AND_ASSIGN(bool mutated, call_inliner.Run(module.get()));
+  absl::StatusOr<bool> filecheck_result = RunFileCheck(module->ToString({}), R"(
+  //CHECK: %lhs.2 = f32[] constant(42)
+  //CHECK: %call1 = f32[] call(f32[] %lhs.2), to_apply=%sub, frontend_attributes={_xla_stream_annotation="1"}
+  //CHECK: %rhs.2 = f32[] constant(2)
+  //CHECK: ROOT %add.1 = f32[] add(f32[] %call1, f32[] %rhs.2)
+  )");
+  TF_ASSERT_OK(filecheck_result.status());
+  EXPECT_TRUE(*filecheck_result);
+
```
Comment (golechwierowicz) [inline xla/service/call_inliner_test.cc:530]: For the future: I think everything below that can be deleted, because it's already tested via filecheck.

**Review by frgossen — APPROVED** [review]

**Pull request closed without merging**